go/types/value_decoder.go

```diff
@@ -8,33 +8,48 @@ import (
 	"fmt"
 
 	"github.com/attic-labs/noms/go/d"
-	"github.com/attic-labs/noms/go/hash"
 )
 
 type valueDecoder struct {
-	nomsReader
+	binaryNomsReader
 	vrw        ValueReadWriter
 	validating bool
 }
 
-// |tc| must be locked as long as the valueDecoder is being used
-func newValueDecoder(nr nomsReader, vrw ValueReadWriter) *valueDecoder {
+func newValueDecoder(buff []byte, vrw ValueReadWriter) *valueDecoder {
+	nr := binaryNomsReader{buff, 0}
```

Comment:
yeah, lets make this a value type and change all of its methods to not take pointer receivers. that should have the heap allocation

Reply:
That does not work since the methods mutate internal state.

Reply:
ah. sorry. i was thinking that binaryNomsReader was an interface reference, but it's an embed. sorry.